### PR TITLE
Implemented DSL graph answerer + clientActor

### PR DESCRIPTION
**Pull request by RaulinN**
- 333e17f : DSL graph answerer + client actor
  - The answerer gives an answer a client (front-end) query
  - The client actor represents a client (front-end) that is currently
  connected to the server

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/Server.scala [hunk 1]
```diff
@@ -25,37 +25,34 @@ object Server {
     val PORT = 8000
     val PATH = ""
 
-    val root = Behaviors.setup[Nothing] { context =>
-      implicit val system: ActorSystem[Nothing] = context.system
+    val system = akka.actor.ActorSystem("magic")
```
Comment (pierluca) [inline be2-scala/src/main/scala/ch/epfl/pop/Server.scala:28]: We should probably use another string :)

Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/Server.scala:29]: Done

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala [hunk 2]
```diff
@@ -1,5 +1,71 @@
...
+      // ActorFlow.ask(clientActorRef)(makeMessage = (el, replyTo: ActorRef) => SubscribeTo(channel))
+      clientActorRef ! ClientActor.SubscribeTo(jsonRpcMessage.channel)
+      Right(PipelineError(-100, "FIXME: should use akka ask pattern to create an ActorFlow"))
+    case graphMessage@_ => graphMessage
+  }
+
+  def unsubscribeHandler(clientActorRef: ActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Left(jsonRpcMessage: Unsubscribe) =>
+      clientActorRef ! ClientActor.UnsubscribeFrom(jsonRpcMessage.channel)
+      Right(PipelineError(-100, "FIXME: should use akka ask pattern to create an ActorFlow"))
```
Comment (gnarula) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala:63]:
```suggestion
      Right(PipelineError(ErrorCodes.SERVER_FAULT.id, "FIXME: should use akka ask pattern to create an ActorFlow"))
```

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala [hunk 3]
```diff
@@ -1,5 +1,71 @@
...
+    }
+  })
+
+  case class Asking(g: GraphMessage, replyTo: ActorRef)
+
+  def subscribeHandler(clientActorRef: ActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Left(jsonRpcMessage: Subscribe) =>
+      // ActorFlow.ask(clientActorRef)(makeMessage = (el, replyTo: ActorRef) => SubscribeTo(channel))
+      clientActorRef ! ClientActor.SubscribeTo(jsonRpcMessage.channel)
+      Right(PipelineError(-100, "FIXME: should use akka ask pattern to create an ActorFlow"))
```
Comment (gnarula) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala:56]:
```suggestion
      Right(PipelineError(ErrorCodes.SERVER_FAULT.id, "FIXME: should use akka ask pattern to create an ActorFlow"))
```

Comment (gnarula) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala:56]: I'm not sure, but is this where we'd handle a Nack?

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/Answerer.scala [hunk 4]
```diff
@@ -1,12 +1,51 @@
...
+      .collect { case graphMessage: GraphMessage => ClientAnswer(graphMessage) }
+      // Send the ClientAnswer to clientActorRef. Whenever the stream between the client
+      // actor and the actual client (front-end) is broken, the message DisconnectWsHandle
+      // is sent to clientActorRef
+      .to(Sink.actorRef(clientActorRef, DisconnectWsHandle))
+
+    // Integration point between Akka Streams and above actor
+    val source: Source[TextMessage, NotUsed] = Source
+      // By using .actorRef, the source emits whatever the actor "wsHandle" sends
+      .actorRef(bufferSize = 10, overflowStrategy = OverflowStrategy.dropBuffer)
```
[the remark this answers is not in the document]
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/Answerer.scala:36]: I upgraded the size of the buffer to 50 for now. Moreover, here are all the possible overflow strategies. I decided to go for the `backpressure` strategy as this is the only one that doesn't drop messages (or fails)

```scala
object OverflowStrategy {

  /**
   * If the buffer is full when a new element arrives, drops the oldest element from the buffer to make space for
   * the new element.
   */
  def dropHead: OverflowStrategy = DropHead(Logging.DebugLevel)

  /**
   * If the buffer is full when a new element arrives, drops the youngest element from the buffer to make space for
   * the new element.
   */
  def dropTail: OverflowStrategy = DropTail(Logging.DebugLevel)

  /**
   * If the buffer is full when a new element arrives, drops all the buffered elements to make space for the new element.
   */
  def dropBuffer: OverflowStrategy = DropBuffer(Logging.DebugLevel)

  /**
   * If the buffer is full when a new element arrives, drops the new element.
   */
  def dropNew: OverflowStrategy = DropNew(Logging.DebugLevel)

  /**
   * If the buffer is full when a new element is available this strategy backpressures the upstream publisher until
   * space becomes available in the buffer.
   */
  def backpressure: OverflowStrategy = Backpressure(Logging.DebugLevel)

  /**
   * If the buffer is full when a new element is available this strategy completes the stream with failure.
   */
  def fail: OverflowStrategy = Fail(Logging.ErrorLevel)
}
```

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala (hunk 2 above)
Comment (RaulinN) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsHandler.scala:63]: This is something that should be modified asap. This is the last problem I have with the pubsub. In theory, there is a way to make a flow out of an actor (with Actorflow on line 54) but I cannot yet manange to make it compile

**Review by gnarula — APPROVED** [review]

**Pull request merged**